### PR TITLE
Create AP in genesis

### LINKED ISSUE #223: Create AP in genesis
Add a new section to the PoE genesis init.

- [ ] AP will have the same standard settings from OC (like escrow amount, voting quorum, etc) and also a few custom ones for the dispute resolution contract. [init-msg](https://github.com/confio/tgrade-contracts/blob/v0.6.2/contracts/tgrade-ap-voting/src/msg.rs#L13)

See https://github.com/confio/tgrade-contracts/issues/415 https://github.com/confio/tgrade-contracts/issues/416 and https://github.com/confio/tgrade-contracts/issues/417

- [ ] In genesis, create a trusted circle like with OC. Then create a voting contract like tgrade-oc-voting but using a new custom contract: [tgrade_ap_voting.wasm](https://github.com/confio/tgrade/blob/main/x/poe/contract/tgrade_ap_voting.wasm)
[embed and store on chain](https://github.com/confio/tgrade/blob/v0.6.2/x/poe/bootstrap.go#L33-L34)

Note: Contract is **not privileged**. 
- [ ] Currently migrator is valset, like all other poe contracts [example](https://github.com/confio/tgrade/blob/v0.6.2/x/poe/bootstrap.go#L247)

- [ ] The addresses of both contracts should be query-able just like OC. Add new [contract type](https://github.com/confio/tgrade/blob/v0.7.0/x/poe/types/poe.pb.go#L36-L48) and store in bootstrap

AP may require a code ID to be uploaded in bootstrap (and set in InstantiateMsg) but never actually instantiated. This would be used for the group of judges to vote on individual cases and instantiated as needed.

### PR DESCRIPTION
**Pull request by pinosu**
Resolves #223 

- [x] AP will have the same standard settings from OC (like escrow amount, voting quorum, etc) and also a few custom ones for the dispute resolution contract. [init-msg](https://github.com/confio/tgrade-contracts/blob/v0.6.2/contracts/tgrade-ap-voting/src/msg.rs#L13)

- [x] In genesis, create a trusted circle like with OC. Then create a voting contract like tgrade-oc-voting but using a new custom contract: [tgrade_ap_voting.wasm](https://github.com/confio/tgrade/blob/main/x/poe/contract/tgrade_ap_voting.wasm)
[embed and store on chain](https://github.com/confio/tgrade/blob/v0.6.2/x/poe/bootstrap.go#L33-L34)

- [x] Currently migrator is valset, like all other poe contracts [example](https://github.com/confio/tgrade/blob/v0.6.2/x/poe/bootstrap.go#L247)

- [x] The addresses of both contracts should be query-able just like OC. Add new [contract type](https://github.com/confio/tgrade/blob/v0.7.0/x/poe/types/poe.pb.go#L36-L48) and store in bootstrap

### CODE REVIEW

File: proto/confio/poe/v1beta1/genesis.proto [hunk 1]
```diff
@@ -187,3 +191,24 @@ message VotingRules {
...
+  // DenyListContractAddress is an optional cw4 contract with list of addresses
+  // denied to be part of TrustedCircle
+  string deny_list_contract_address = 4;
+
+  // DisputeCost The required dispute amount, in the default denom (utgd)
+  cosmos.base.v1beta1.Coin dispute_cost = 5 [ (gogoproto.nullable) = false ];
+
+  google.protobuf.Duration waiting_period = 6
+      [ (gogoproto.nullable) = false, (gogoproto.stdduration) = true ];
+}
```
Comment (alpe) [inline proto/confio/poe/v1beta1/genesis.proto:214]: looks good

File: proto/confio/poe/v1beta1/poe.proto [hunk 2]
```diff
@@ -34,6 +34,8 @@ enum PoEContractType {
       [ (gogoproto.enumvalue_customname) = "PoEContractTypeCommunityPool" ];
   VALIDATOR_VOTING = 9
       [ (gogoproto.enumvalue_customname) = "PoEContractTypeValidatorVoting" ];
+  ARBITER_POOL = 10
```
Comment (alpe) [inline proto/confio/poe/v1beta1/poe.proto:37]: 👍

File: testing/smoke_test.go [hunk 3]
```diff
@@ -36,8 +36,8 @@ func TestSmokeTest(t *testing.T) {
 	codes := gjson.Get(qResult, "code_infos.#.code_id").Array()
 	t.Log("got query result", qResult)
 
-	const poeContractCount = 9
-	const poeCodeCount = 8
+	const poeContractCount = 10
+	const poeCodeCount = 9
```
Comment (alpe) [inline testing/smoke_test.go:40]: 👍

File: x/poe/bootstrap.go [hunk 4]
```diff
@@ -295,6 +306,34 @@ func BootstrapPoEContracts(ctx sdk.Context, k wasmtypes.ContractOpsKeeper, tk tw
 	return nil
 }
 
+func createTrustedCircle(ctx sdk.Context, contractAddr sdk.AccAddress, tk twasmKeeper, members []string, sender sdk.AccAddress, deposit sdk.Coin) error {
```
Comment (alpe) [inline x/poe/bootstrap.go:309]: nit: can be `addToTrustedCircle` as the TC exists already

**Review by alpe — COMMENTED** [review]
nice start

File: x/poe/bootstrap.go [hunk 5]
```diff
@@ -280,6 +263,34 @@ func BootstrapPoEContracts(ctx sdk.Context, k wasmtypes.ContractOpsKeeper, tk tw
...
+	}
+
+	// setup arbiter pool
+	apCodeID, err := k.Create(ctx, systemAdminAddr, tgArbiterPool, &wasmtypes.AllowEverybody)
+	if err != nil {
+		return sdkerrors.Wrap(err, "store tg ap contract: ")
+	}
+
+	logger.Info("arbiter pool contract stored", "code_id", apCodeID)
+
```
Comment (alpe) [inline x/poe/bootstrap.go:293]: Please pass the `apContractAddr` and the new `apVotingContractAddr` to the `setAllPoEContractsInstanceMigrators` call in L297

File: x/poe/bootstrap.go [hunk 6]
```diff
@@ -280,6 +263,34 @@ func BootstrapPoEContracts(ctx sdk.Context, k wasmtypes.ContractOpsKeeper, tk tw
...
+		if err != nil {
+			return err
+		}
+	}
+
+	// setup arbiter pool
+	apCodeID, err := k.Create(ctx, systemAdminAddr, tgArbiterPool, &wasmtypes.AllowEverybody)
+	if err != nil {
+		return sdkerrors.Wrap(err, "store tg ap contract: ")
+	}
```
Comment (alpe) [inline x/poe/bootstrap.go:290]: Please 
- instantiate the arbiter pool contract with this [init message](https://github.com/confio/tgrade-contracts/blob/v0.6.2/contracts/tgrade-ap-voting/src/msg.rs#L13) and 
- store the contract address with a new contract type `PoEContractTypeArbiterPoolVoting`

File: x/poe/bootstrap.go [hunk 7]
```diff
@@ -342,6 +381,34 @@ func newOCGovProposalsInitMsg(gs types.GenesisState, ocContract, engagementContr
...
+	return contract.TrustedCircleInitMsg{
+		Name:                      cfg.Name,
+		EscrowAmount:              cfg.EscrowAmount.Amount,
+		VotingPeriod:              cfg.VotingRules.VotingPeriod,
+		Quorum:                    *contract.DecimalFromPercentage(cfg.VotingRules.Quorum),
+		Threshold:                 *contract.DecimalFromPercentage(cfg.VotingRules.Threshold),
+		AllowEndEarly:             cfg.VotingRules.AllowEndEarly,
+		InitialMembers:            []string{}, // sender is added to AP by default in the contract
+		DenyList:                  cfg.DenyListContractAddress,
+		EditTrustedCircleDisabled: true,
```
Comment (alpe) [inline x/poe/bootstrap.go:396]: This flag is copied from OC setup as described in the issue. Just want to double check with @ethanfrey if this is applies here, too.

File: x/poe/bootstrap_integration_test.go [hunk 8]
```diff
@@ -56,12 +56,17 @@ func setupPoEContracts(t *testing.T, mutators ...func(m *types.GenesisState)) (s
...
+		for _, member := range members {
+			addr, err := sdk.AccAddressFromBech32(member)
+			require.NoError(t, err)
+			example.Faucet.Fund(ctx, addr, sdk.NewCoin(types.DefaultBondDenom, coins))
+		}
 	}
 
+	fundMembers(gs.OversightCommunityMembers, sdk.NewInt(1_000_000))
+	fundMembers(gs.ArbiterPoolMembers, sdk.NewInt(1_000_000))
+
```
Comment (alpe) [inline x/poe/bootstrap_integration_test.go:63]: Let's check the AP members setup in this test

File: x/poe/contract/common_test.go [hunk 9]
```diff
@@ -33,12 +33,17 @@ func setupPoEContracts(t *testing.T, mutators ...func(m *types.GenesisState)) (s
...
+	fundMembers := func(members []string, coins sdk.Int) {
+		for _, member := range members {
+			addr, err := sdk.AccAddressFromBech32(member)
+			require.NoError(t, err)
+			example.Faucet.Fund(ctx, addr, sdk.NewCoin(types.DefaultBondDenom, coins))
+		}
 	}
 
+	fundMembers(gs.OversightCommunityMembers, sdk.NewInt(1_000_000))
+	fundMembers(gs.ArbiterPoolMembers, sdk.NewInt(1_000_000))
```
Comment (alpe) [inline x/poe/contract/common_test.go:45]: 👍

File: x/poe/types/test_fixtures.go [hunk 10]
```diff
@@ -41,6 +41,7 @@ func GenesisStateFixture(mutators ...func(m *GenesisState)) GenesisState {
 	}}
 
 	r.OversightCommunityMembers = []string{RandomAccAddress().String(), RandomAccAddress().String()}
+	r.ArbiterPoolMembers = []string{RandomAccAddress().String(), RandomAccAddress().String()}
```
Comment (alpe) [inline x/poe/types/test_fixtures.go:44]: 🥇

File: x/poe/bootstrap.go (hunk 5 above)
Comment (alpe) [reply x/poe/bootstrap.go:301]: Sorry, this is not correct. All PoE contracts are fetched by `types.PoEContractType_value` so this is done already.

File: x/poe/bootstrap_integration_test.go [hunk 11]
```diff
@@ -127,3 +126,20 @@ func withRandomValidators(t *testing.T, ctx sdk.Context, example keeper.TestKeep
...
+	}
+	notVotingMembers := make(map[string]struct{}, len(members))
+	for _, m := range members {
+		notVotingMembers[m] = struct{}{}
+	}
+
+	for _, v := range voters.Members {
+		delete(notVotingMembers, v.Addr)
+	}
+	return len(notVotingMembers) == 0, nil
```
Comment (alpe) [inline x/poe/bootstrap_integration_test.go:144]: There is still the case that the contract contains more members than you have passed. Please ensure that len(voters)==len(members)

File: x/poe/bootstrap.go [hunk 12]
```diff
@@ -295,6 +314,34 @@ func BootstrapPoEContracts(ctx sdk.Context, k wasmtypes.ContractOpsKeeper, tk tw
 	return nil
 }
 
+func addToTrustedCircle(ctx sdk.Context, contractAddr sdk.AccAddress, tk twasmKeeper, members []string, sender sdk.AccAddress, deposit sdk.Coin) error {
```
Comment (alpe) [inline x/poe/bootstrap.go:317]: good helper method!

File: x/poe/bootstrap.go [hunk 13]
```diff
@@ -280,6 +263,42 @@ func BootstrapPoEContracts(ctx sdk.Context, k wasmtypes.ContractOpsKeeper, tk tw
...
 
+	// setup trusted circle for ap
+	apTrustedCircleInitMsg := newAPTrustedCircleInitMsg(gs)
+	apDeposit := sdk.NewCoins(gs.ArbiterPoolContractConfig.EscrowAmount)
+	firstAPMember, err := sdk.AccAddressFromBech32(gs.ArbiterPoolMembers[0])
+	if err != nil {
+		return sdkerrors.Wrap(err, "first ap member")
+	}
+
+	apContractAddr, _, err := k.Instantiate(ctx, trustedCircleCodeID, firstAPMember, systemAdminAddr, mustMarshalJson(apTrustedCircleInitMsg), "arbter_pool", apDeposit)
```
Comment (alpe) [inline x/poe/bootstrap.go:274]: typo:
```suggestion
	apContractAddr, _, err := k.Instantiate(ctx, trustedCircleCodeID, firstAPMember, systemAdminAddr, mustMarshalJson(apTrustedCircleInitMsg), "arbiter_pool", apDeposit)
```

File: x/poe/bootstrap.go [hunk 14]
```diff
@@ -280,6 +263,42 @@ func BootstrapPoEContracts(ctx sdk.Context, k wasmtypes.ContractOpsKeeper, tk tw
...
+		err = addToTrustedCircle(ctx, apContractAddr, tk, gs.ArbiterPoolMembers[1:], firstAPMember, gs.ArbiterPoolContractConfig.EscrowAmount)
+		if err != nil {
+			return err
+		}
+	}
+
+	// setup arbiter pool
+	apCodeID, err := k.Create(ctx, systemAdminAddr, tgArbiterPool, &wasmtypes.AllowEverybody)
+	if err != nil {
+		return sdkerrors.Wrap(err, "store tg ap contract: ")
```
Comment (alpe) [inline x/poe/bootstrap.go:289]:
```suggestion
		return sdkerrors.Wrap(err, "store arbiter voting contract: ")
```

File: x/poe/bootstrap_integration_test.go [hunk 15]
```diff
@@ -24,47 +24,46 @@ import (
...
-
-	for _, v := range voters.Members {
-		delete(notVotingMembers, v.Addr)
-	}
-	assert.Empty(t, notVotingMembers)
+	apContractAddr, err := example.PoEKeeper.GetPoEContractAddress(ctx, types.PoEContractTypeArbiterPool)
+	require.NoError(t, err)
+	areAllVotingMembers, err = membersAreAllVotingMembers(ctx, apMembers, apContractAddr, example.TWasmKeeper)
+	require.NoError(t, err)
+	assert.True(t, areAllVotingMembers)
```
Comment (alpe) [inline x/poe/bootstrap_integration_test.go:39]: You can check here that code_ids are pinned and none of the contract instances are privileged to cover acceptance criteria and have some regression tests.

**Review by alpe — APPROVED** [review]
Really good work! 
I added some nits and comments to improve test coverage.

File: x/poe/bootstrap.go (hunk 7 above)
Comment (kmw101) [reply x/poe/bootstrap.go:404]: This is correct we do not want the AP to be able to edit the TC.

**Comment by alpe** [comment]
Good work. Thank you 💐

**Pull request merged**